mem-store.ts

```diff
@@ -106,18 +107,129 @@ function mem_store(options: any) {
         if (shouldMerge) {
           mement = Object.assign(prev || {}, mement)
         }
-        prev = entmap[base][name][mement.id] = mement
 
-        seneca.log.debug(function () {
-          return [
-            'save/' + (create ? 'insert' : 'update'),
-            ent.canon$({ string: 1 }),
-            mement,
-            desc,
-          ]
+        return upsertIfRequested(msg, function (err: Error, ctx: UpsertIfRequestedContext) {
+          if (err) {
+            return reply(err);
+          }
+
+          const { did_upsert } = ctx;
+
+          if (did_upsert) {
+            const { out } = ctx;
+            return reply(null, out);
+          }
+
+
+          prev = entmap[base][name][mement.id] = mement;
+
+          seneca.log.debug(function () {
+            return [
+              'save/' + (isNewEntityBeingCreated(msg) ? 'insert' : 'update'),
+              ent.canon$({ string: 1 }),
+              mement,
+              desc,
+            ];
+          });
+
+          return reply(null, ent.make$(prev));
         })
 
-        reply(null, ent.make$(prev))
+
+        type UpsertIfRequestedContext = { did_upsert: boolean, out?: any }
+
+        type UpsertIfRequestedCallback = (
+          err:     Error | null,
+          result?: UpsertIfRequestedContext 
+        ) => any
+
+        function upsertIfRequested(msg: any, reply: UpsertIfRequestedCallback) {
+          // This is the query passed to the .save$ method.
+          //
+          const query_for_save = msg.q;
+
+
+          if (isNewEntityBeingCreated(msg) && Array.isArray(query_for_save.upsert$)) {
+            const upsert_on = query_for_save.upsert$
+              //
+              // NOTE: Here we are stripping private properties from the
+              // upsert$ fields. That's because the underlying `listents`
+              // helper ignores private properties completely, - in ways
+              // that lead to situations where:
+              // ```
+              //  .list$({ i_am_private_prop$: 'abc' })
+              // ```
+              // - returns a list of all entities.
+              //
+              .filter((p: string) => !isPrivateProp(p));
+
+
+            if (upsert_on.length > 0) {
+              const qent = msg.ent;
+              const public_entdata = msg.ent.data$(false);
+
+              const q = upsert_on.reduce((acc: any, upsert_on_field: string) => {
+                acc[upsert_on_field] = public_entdata[upsert_on_field];
+                return acc;
+              }, {});
+
+
+              return listents(seneca, entmap, qent, q, function (err: Error | null, docs_to_update: any[]) {
```

Comment:
@rjrodger 
The implementation of the `load` method relies on the `listents` helper. Would we prefer the suggested simpler approach - with the `Array#find` suggestion, or a more "idiomatic" one in terms of the code already present in the module - with the usage of the `listents` helper?

Reply:
load uses listents because it has to support complex queries (gt, lt, etc are possible), but upsert is much simpler - it must be an *exact* match, so you use find directly - this saves a callback level

Reply:
Addressed:
https://github.com/senecajs/seneca-mem-store/blob/c9a4126230d2b3fb25adc6c51ab5ce5c13438914/mem-store.ts#L172